engine/src/cmd/missile.cpp

```diff
@@ -0,0 +1,364 @@
+#include "unit_util.h"
+#include "vsfilesystem.h"
+#include "star_system.h"
+#include "universe.h"
+#include "dummy_unit.h"
+#include "game_config.h"
+
+
+////////////////////////////////////////////////////////////////
+/// MissileEffect
+////////////////////////////////////////////////////////////////
+MissileEffect::MissileEffect( const QVector &pos, float dam, float pdam, float radius, float radmult, void *owner ): pos( pos )
+{
+    damage = dam;
+    phasedamage = pdam;
+    this->radius     = radius;
+    radialmultiplier = radmult;
+    this->ownerDoNotDereference = owner;
+}
+
+void MissileEffect::ApplyDamage( Unit *smaller )
+{
+    QVector norm = pos-smaller->Position();
+    float smaller_rsize = smaller->rSize();
+    float distance = norm.Magnitude()-smaller_rsize;            // no better check than the bounding sphere for now
+    if ( distance < radius) {                                   // "smaller->isUnit() != MISSILEPTR &&" was removed - why disable antimissiles?
+        if ( distance < 0)
+            distance = 0.f;                                     //it's inside the bounding sphere, so we'll not reduce the effect
+        if (radialmultiplier < .001) radialmultiplier = .001;
+        float dist_part=distance/radialmultiplier;              //radialmultiplier is radius of the set damage
+        float damage_mul;
+        if ( dist_part > 1.f) {                                  // there can be something else, such as different eye- and ear- candy
+            damage_mul = 1/(dist_part*dist_part);
+        }
+        else {
+            damage_mul = 2.f - dist_part*dist_part;
+        }
+        /*
+         *  contrived formula to create paraboloid falloff rather than quadratic peaking at 2x damage at origin
+         *  k = 2-distance^2/radmul^2
+         *
+         * if the explosion itself was a weapon, it would have double the base damage, longrange=0.5 (counted at Rm) and more generic form:
+         * Kclose = 1-(1-longrange)*(R/Rm)^2
+         * Kfar   = longrange/(R/Rm)^2
+         * or Kapprox = longrange/(longrange-(R/Rm)^3*(1-longrange)) ; obviously, with more checks preventing /0
+         */
+        DoApplyDamage(smaller, smaller, distance, damage_mul);
+    }
+}
+
+const QVector&  MissileEffect::GetCenter() const
+{
+    return pos;
+}
+
+
+float  MissileEffect::GetRadius() const
+{
+    return radius;
+}
+
+
+void MissileEffect::DoApplyDamage(Unit *parent, Unit *un, float distance, float damage_fraction) {
+    QVector norm = pos-un->Position();
+    norm.Normalize();
+    float damage_left = 1.f;
+    if (un->hasSubUnits()) {
+        /*
+         * Compute damage aspect ratio of each subunit with their apparent size ( (radius/distance)^2 )
+         * and spread damage across affected subunits based on their apparent size vs total spread surface
+         */
+        double total_area = 0.0f; {
+            un_kiter ki = un->viewSubUnits();
+            for (const Unit *subun; (subun = *ki); ++ki) {
```

Comment:
Just for my understanding here... this is an infinite loop, right? 
The middle bit of the for loop that is supposed to do checking is doing assigning with the single "="... right?

So, this loop exists with either the enemy killed, or the sphere of influence surpassed?

Reply:
@evertvorster  no, it's not an infinite loop, but it's not obvious what's happening either.
```c++
un_kiter ki = un->viewSubUnits();  // creates the variable `ki` which points to a sub unit in some form, it's not obvious whether this is a pointer or not.

for (
   const Unit* subun;  // this just declares a type of `const Unit*`
   (subun = *ki);  // this is the evaluation, but also assigns `subun` to the what is pointed to by dereferencing `ki`; this no-obviously terminates the loop when `subun` and `*ki` by extension are both NULL pointers and it must be a `0` value.
  ++ki // pointer math to move to the next sub-unit
)
```
The evaluation also doesn't verify that `ki` is still pointing to something valid to dereference. The code implies that `ki` is a double-pointer (pointer to a pointer to a Unit).

This would be better written as:
```
for (auto ki = un->viewSubUnits(); *ki != NULL; ++ki) {
    // implies that ki is a pointer to a pointer, so needs to be dereferenced
    const Unit* subun = *ki;
   // ensure that subun is still a valid object to point to
    if (subun == NULL) break;
    ...
}
```

Reply:
Thanks so much for the explanation, learn something every day!

Reply:
I thought that ki was an iterator; that dereferencing it meant retrieving the value that the iterator is currently pointing to; and that `++ki` meant advancing the iterator. In which case, no pointer math involved.

Reply:
@stephengtuggy yes, that's likely the case - I don't know the type behind `un_kiter` so was just going off of what was incontext. Overall, the same thing applies though with the slight adjustment for checking for the end of the iteration instead of a NULL/Zero/False value. It should look more like the following:
```
std::vector<int> a;
for (auto x = a.begin(); x != a.end(); ++a) {...}
```

Reason being is that the end of an iteration doesn't necessarily translate to a false/zero value - it's a valid object in its own right, and one that doesn't equate to NULL, zero, or False. It may be that `un_kiter` is a pseudo-iterator, in which case we should change it to be a proper iterator, but that's beyond the scope of this PR.

In any respect, the `for` loop is constructed incorrectly. It may work, but that's likely by-chance and not by-design and certainly not straight forward in its behavior.

Reply:
@BenjamenMeyer Agreed.

Reply:
Agree as well. However, I can't touch every line of code.  This was simply copied from another file as is. It looks like code that iterates over subunits in range and damages/kills them.

Reply:
@royfalk that's fine right now - it's something we need to come back and address over the whole code base.